### PR TITLE
Projection documentation

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
Towards #1209

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            {
+                new DataPoint(){ Features = new float[4] { 1, 1, 0, 0} },
+                new DataPoint(){ Features = new float[4] { 2, 2, 0, 0} },
+                new DataPoint(){ Features = new float[4] { 1, 0, 1, 0} },
+                new DataPoint(){ Features = new float[4] { 0, 1, 0, 1} }
+            };
+            // Convert training data to IDataView, the general data type used in ML.NET.
+            var data = mlContext.Data.LoadFromEnumerable(samples);
+            // NormalizeLpNorm normalize rows individually by rescaling them to unit norm.
+            // Performs the following operaion on a row X:  Y = (X - M) / D where M is mean, and D is selected norm.
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs]: What is the selected norm? Is it norm of the feature vector in a row being processed? Also, what are the shapes of X, Y, M, and D?

Comment (Ivanidzo4ka) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs]: `mean` -> `mean vector` 
`D is selected norm` -> `D is calculated value of selected norm parameter` 
Does that sound better?

---
In reply to: [273152443](https://github.com/dotnet/machinelearning/pull/3232#discussion_r273152443) [](ancestors = 273152443)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ApproximatedKernelMap.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+using System.Linq;
+using Microsoft.ML;
+using Microsoft.ML.Data;
+using Microsoft.ML.Transforms;
+
+namespace Samples.Dynamic
+{
+    public static class ApproximatedKernelMap
+    {
+        public static void Example()
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ApproximatedKernelMap.cs]:
```suggestion
        // Transform feature vector to another non-linear space. See https://people.eecs.berkeley.edu/~brecht/papers/07.rah.rec.nips.pdf.
        public static void Example()
```

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeGlobalContrast.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            // 2.0000, 2.0000,-2.0000,-2.0000
+            // 2.0000,-2.0000, 2.0000,-2.0000
+            //-2.0000, 2.0000,-2.0000, 2.0000
+            foreach (var row in column)
+                Console.WriteLine(string.Join(", ", row.Select(x => x.ToString("f4"))));
+        }
+
+        private class DataPoint
+        {
+            [VectorType(7)]
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeGlobalContrast.cs]: >7 [](start = 24, length = 1)

It shouldn't work!

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            // 0.2500,  0.2500, -0.2500, -0.2500
+            // 0.2500, -0.2500,  0.2500, -0.2500
+            //-0.2500,  0.2500, -0.2500,  0.2500
+            foreach (var row in column)
+                Console.WriteLine(string.Join(", ", row.Select(x => x.ToString("f4"))));
+        }
+
+        private class DataPoint
+        {
+            [VectorType(7)]
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs]: it shouldn't work!

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs (hunk 1 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs]: Norm on what? A column? A row? Is it a scalar? In tensor computation, norm operation can produce another tensor. 

Say, if I have rows, x_1, x_2, x_3. Is D=1/3 (x_1 + x_2 + x_3) true? Or D=ReduceSum(x_i) for the x subscripted by i? In addition, is D=||x_i||_2 for the x subscripted by i?

---
In reply to: [273153905](https://github.com/dotnet/machinelearning/pull/3232#discussion_r273153905) [](ancestors = 273153905,273152443)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ApproximatedKernelMap.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            // Now we can transform the data and look at the output to confirm the behavior of the estimator.
+            // This operation doesn't actually evaluate data until we read the data below.
+            var tansformer = approximation.Fit(data);
+            var transformedData = tansformer.Transform(data);
+
+            var column = transformedData.GetColumn<float[]>("Features").ToArray();
+            //-0.0119, 0.5867, 0.4942,  0.7041
+            // 0.4720, 0.5639, 0.4346,  0.2671
+            //-0.2243, 0.7071, 0.7053, -0.1681
+            // 0.0846, 0.5836, 0.6575,  0.0581
```
Comment (artidoro) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ApproximatedKernelMap.cs]: Could you add these lines below after using:
// Expected output:

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs [hunk 6]
```diff
@@ -279,7 +279,7 @@ internal static LpNormNormalizingEstimator NormalizeLpNorm(this TransformsCatalo
         /// <example>
         /// <format type="text/markdown">
         /// <![CDATA[
-        /// [!code-csharp[GlobalContrastNormalize](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs?range=1-6,12-112)]
+        /// [!code-csharp[GlobalContrastNormalize](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeGlobalContrast.cs)]
```
Comment (rogancarr) [inline src/Microsoft.ML.Transforms/NormalizerCatalog.cs]: >GlobalContrastNormalize [](start = 26, length = 23)

NormalizeGlobalContrast

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs [hunk 7]
```diff
@@ -249,7 +249,7 @@ public static class NormalizationCatalog
         /// <example>
         /// <format type="text/markdown">
         /// <![CDATA[
-        /// [!code-csharp[LpNormalize](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs?range=1-6,12-112)]
+        /// [!code-csharp[LpNormalize](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs)]
```
Comment (rogancarr) [inline src/Microsoft.ML.Transforms/NormalizerCatalog.cs]: >LpNormalize [](start = 26, length = 11)

NormalizeLpNorm

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                new DataPoint(){ Features = new float[4] { 1, 0, 1, 0} },
+                new DataPoint(){ Features = new float[4] { 0, 1, 0, 1} }
+            };
+            // Convert training data to IDataView, the general data type used in ML.NET.
+            var data = mlContext.Data.LoadFromEnumerable(samples);
+            // NormalizeLpNorm normalize rows individually by rescaling them to unit norm.
+            // Performs the following operaion on a row X:  Y = (X - M(X)) / D(X) 
+            // where M(X) is scalar value of mean for current row,
+            // and D(X) is scalar value of selected `norm` parameter .
+            var approximation = mlContext.Transforms.NormalizeLpNorm("Features", norm: LpNormNormalizingEstimatorBase.NormFunction.L1, ensureZeroMean: true);
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs]: >ensureZeroMean [](start = 135, length = 14)

What does EnsureZeroMean do? Subtract the mean?

File: src/Microsoft.ML.Transforms/KernelCatalog.cs [hunk 9]
```diff
@@ -26,7 +26,7 @@ public static class KernelExpansionCatalog
         /// <example>
         /// <format type="text/markdown">
         /// <![CDATA[
-        /// [!code-csharp[CreateRandomFourierFeatures](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs?range=1-6,12-112)]
+        /// [!code-csharp[CreateRandomFourierFeatures](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ApproximatedKernelMap.cs)]
```
Comment (rogancarr) [inline src/Microsoft.ML.Transforms/KernelCatalog.cs]: >CreateRandomFourierFeatures [](start = 26, length = 27)

ApproximatedKernelMap

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ApproximatedKernelMap.cs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            // Now we can transform the data and look at the output to confirm the behavior of the estimator.
+            // This operation doesn't actually evaluate data until we read the data below.
+            var tansformer = approximation.Fit(data);
+            var transformedData = tansformer.Transform(data);
+
+            var column = transformedData.GetColumn<float[]>("Features").ToArray();
+            foreach (var row in column)
+                Console.WriteLine(string.Join(", ", row.Select(x => x.ToString("f4"))));
+            // Expected output:
+            // -0.0119, 0.5867, 0.4942,  0.7041
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ApproximatedKernelMap.cs]: >  [](start = 14, length = 1)

Space Space

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ApproximatedKernelMap.cs [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            //  0.4720, 0.5639, 0.4346,  0.2671
+            // -0.2243, 0.7071, 0.7053, -0.1681
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ApproximatedKernelMap.cs]: >-0.1681 [](start = 40, length = 7)

Seems a bit high?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeGlobalContrast.cs [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+                new DataPoint(){ Features = new float[4] { 2, 2, 0, 0} },
+                new DataPoint(){ Features = new float[4] { 1, 0, 1, 0} },
+                new DataPoint(){ Features = new float[4] { 0, 1, 0, 1} }
+            };
+            // Convert training data to IDataView, the general data type used in ML.NET.
+            var data = mlContext.Data.LoadFromEnumerable(samples);
+            // NormalizeLpNorm normalize rows individually by rescaling them to unit norm.
+            // Performs the following operaion on a row X:  Y = scale *(X - M(X)) / D(X)
+            // where M(X) is scalar value of mean for current row,
+            // and D(X) is scalar value of either Standard deviation or L2 norm.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeGlobalContrast.cs]: This comment looks like a copy/paste holdover.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ApproximatedKernelMap.cs (hunk 10 above)
Comment (Ivanidzo4ka) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ApproximatedKernelMap.cs]: I prefer to align numbers, so one space was taken by minus sign.

---
In reply to: [273740487](https://github.com/dotnet/machinelearning/pull/3232#discussion_r273740487) [](ancestors = 273740487)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeGlobalContrast.cs (hunk 12 above)
Comment (Ivanidzo4ka) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeGlobalContrast.cs]: Can you come up with better one?

---
In reply to: [273740748](https://github.com/dotnet/machinelearning/pull/3232#discussion_r273740748) [](ancestors = 273740748)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs (hunk 8 above)
Comment (Ivanidzo4ka) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs]: yes, added it to comment above.

---
In reply to: [273740225](https://github.com/dotnet/machinelearning/pull/3232#discussion_r273740225) [](ancestors = 273740225)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeGlobalContrast.cs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+            var samples = new List<DataPoint>()
+            {
+                new DataPoint(){ Features = new float[4] { 1, 1, 0, 0} },
+                new DataPoint(){ Features = new float[4] { 2, 2, 0, 0} },
+                new DataPoint(){ Features = new float[4] { 1, 0, 1, 0} },
+                new DataPoint(){ Features = new float[4] { 0, 1, 0, 1} }
+            };
+            // Convert training data to IDataView, the general data type used in ML.NET.
+            var data = mlContext.Data.LoadFromEnumerable(samples);
+            // NormalizeLpNorm normalize rows individually by rescaling them to unit norm.
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeGlobalContrast.cs]: >NormalizeLpNorm [](start = 15, length = 15)

old name?

Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeGlobalContrast.cs]: >normalize [](start = 31, length = 9)

normalize*s*

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeGlobalContrast.cs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+            // Performs the following operaion on a row X:  Y = scale *(X - M(X)) / D(X)
+            // where M(X) is scalar value of mean for current row if ensureZeroMean = true or 0 othewise
+            // and D(X) is scalar value of either Standard deviation or L2 norm.
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeGlobalContrast.cs]: let's actually drop details descriptions. such details belong to the <remarks> section and we don't want to repeat them again here.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs (hunk 1 above)
Comment (shmoradims) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs]: Same here. Please move the final answer to what NormalizeLpNorm does' to the <remarks> section for the estimator.

---
In reply to: [273181155](https://github.com/dotnet/machinelearning/pull/3232#discussion_r273181155) [](ancestors = 273181155,273153905,273152443)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs (hunk 8 above)
Comment (shmoradims) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLpNorm.cs]: Let's move parameter details to xml docstring.

---
In reply to: [273741392](https://github.com/dotnet/machinelearning/pull/3232#discussion_r273741392) [](ancestors = 273741392,273740225)

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs [hunk 15]
```diff
@@ -276,10 +281,16 @@ internal static LpNormNormalizingEstimator NormalizeLpNorm(this TransformsCatalo
         /// <param name="ensureZeroMean">If <see langword="true"/>, subtract mean from each value before normalizing and use the raw input otherwise.</param>
         /// <param name="ensureUnitStandardDeviation">If <see langword="true"/>, resulted vector's standard deviation would be one. Otherwise, resulted vector's L2-norm would be one.</param>
         /// <param name="scale">Scale features by this value.</param>
+        /// <remarks>
+        /// This transform performs the following operation on a row X: Y = scale * (X - M(X)) / D(X)
+        /// where M(X) is scalar value of mean for current row if <paramref name="ensureZeroMean"/>set to <see langword="true"/> or <value>0</value> othewise
```
Comment (wschin) [inline src/Microsoft.ML.Transforms/NormalizerCatalog.cs]:
```suggestion
        /// where M(X) is scalar value of mean for all elements in the current row if <paramref name="ensureZeroMean"/>set to <see langword="true"/> or <value>0</value> othewise
```

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs [hunk 16]
```diff
@@ -276,10 +281,16 @@ internal static LpNormNormalizingEstimator NormalizeLpNorm(this TransformsCatalo
...
+        /// D(X) is scalar value of standard deviation for row if <paramref name="ensureUnitStandardDeviation"/> set to <see langword="true"/> or
+        /// L2 norm value for this row if it set to <see langword="false"/> and scale is <paramref name="scale"/>.
```
Comment (wschin) [inline src/Microsoft.ML.Transforms/NormalizerCatalog.cs]:
```suggestion
        /// L2 norm of this row vector if it set to <see langword="false"/> and scale is <paramref name="scale"/>.
```

**Pull request merged**